drv/sidecar-front-io/src/transceivers.rs

```diff
@@ -0,0 +1,201 @@
+        }
+        if (mask & 0xFFFF0000) > 0 {
+            self.fpgas[1].write(op, addr, [m[3], m[2]])?;
+        }
+
+        Ok(())
+    }
+
+    pub fn set_power_enable(&self, mask: u32) -> Result<(), FpgaError> {
+        self.masked_op(WriteOp::BitSet, mask, Addr::QSFP_CTRL_EN_H)
+    }
+
+    pub fn clear_power_enable(&self, mask: u32) -> Result<(), FpgaError> {
+        self.masked_op(WriteOp::BitClear, mask, Addr::QSFP_CTRL_EN_H)
+    }
+
+    pub fn set_reset(&self, mask: u32) -> Result<(), FpgaError> {
+        self.masked_op(WriteOp::BitSet, mask, Addr::QSFP_CTRL_RESET_H)
+    }
+
+    pub fn clear_reset(&self, mask: u32) -> Result<(), FpgaError> {
+        self.masked_op(WriteOp::BitClear, mask, Addr::QSFP_CTRL_RESET_H)
+    }
+
+    pub fn set_lpmode(&self, mask: u32) -> Result<(), FpgaError> {
+        self.masked_op(WriteOp::BitSet, mask, Addr::QSFP_CTRL_LPMODE_H)
+    }
+
+    pub fn clear_lpmode(&self, mask: u32) -> Result<(), FpgaError> {
+        self.masked_op(WriteOp::BitClear, mask, Addr::QSFP_CTRL_LPMODE_H)
+    }
+
+    pub fn setup_i2c_op(
+        &self,
+        is_read: bool,
+        reg: u8,
+        num_bytes: u8,
+        mask: u32,
+    ) -> Result<(), FpgaError> {
+        let m = mask.as_bytes();
+
+        let i2c_op = if is_read {
+            // Defaulting to RandomRead, rather than Read, because RandomRead
+            // sets the reg addr in the target device, then issues a restart to
+            // do the read at that reg addr. On the other hand, Read just starts
+            // a read wherever the reg addr is after the last transaction.
+            TransceiverI2COperation::RandomRead
+        } else {
+            TransceiverI2COperation::Write
+        };
+
+        if (mask & 0xFFFF) > 0 {
+            let request = TransceiversI2CRequest {
+                reg,
+                num_bytes,
+                mask: [m[1], m[0]],
+                op: ((i2c_op as u8) << 1) | Reg::QSFP::I2C_CTRL::START,
+            };
+
+            self.fpgas[0].write(
+                WriteOp::Write,
+                Addr::QSFP_I2C_REG_ADDR,
+                request,
+            )?;
+        }
+
+        if (mask & 0xFFFF0000) > 0 {
+            let request = TransceiversI2CRequest {
+                reg,
+                num_bytes,
+                mask: [m[3], m[2]],
+                op: ((i2c_op as u8) << 1) | Reg::QSFP::I2C_CTRL::START,
+            };
+            self.fpgas[1].write(
+                WriteOp::Write,
+                Addr::QSFP_I2C_REG_ADDR,
+                request,
+            )?;
+        }
+
+        Ok(())
+    }
+
+    pub fn get_i2c_read_buffer(
+        &self,
+        port: u8,
+        buf: &mut [u8],
+    ) -> Result<(), FpgaError> {
+        let fpga_idx: usize = if port < 16 { 0 } else { 1 };
+        self.fpgas[fpga_idx].read_bytes(Self::read_buffer_address(port), buf)
+    }
+
+    pub fn set_i2c_write_buffer(&self, buf: &[u8]) -> Result<(), FpgaError> {
```

Comment:
The asymmetry here is confusing: why don't we need to specify a `port` when setting the write buffer?

(I'm assuming it's because we always write the same buffer, but it would be good to clarify with a comment)

Reply:
You are correct. I will add a comment explaining that.